### PR TITLE
Compact segment: part 2

### PR DESCRIPTION
**Pull request by small-turtle-1**
### What problem does this PR solve?

1. Add sql command for compaction to trigger it manually.
2. Fix bug of using deprecate segment in build index.
3. Create index for compacted segments.
4. Add more unit test and sql test.

TODO:
1. run compaction task periodically.
2. add delete bitmap in knn scan.
3. clean up deprecate item.
4. better compact algorithm.

Issue link:
https://github.com/infiniflow/infinity/issues/485
### Code changes

- [x] Has Code change
- [ ] Has CI related scripts change

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (knn performance test)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Note for reviewer

### CODE REVIEW

File: src/storage/txn/txn.cpp [hunk 1]
```diff
@@ -295,7 +296,10 @@ Status Txn::DropTableCollectionByName(const String &db_name, const String &table
     return Status::OK();
 }
 
-Status Txn::CreateIndex(TableEntry *table_entry, const SharedPtr<IndexDef> &index_def, ConflictType conflict_type, bool prepare) {
+Status Txn::CreateIndexDef(TableEntry *table_entry,
```
Comment (JinHai-CN) [inline src/storage/txn/txn.cpp:299]: return value: Tuple<TableIndexEntry*, Status>

File: src/executor/operator/physical_command.cppm [hunk 2]
```diff
@@ -54,6 +55,8 @@ public:
     inline SharedPtr<Vector<SharedPtr<DataType>>> GetOutputTypes() const override { return output_types_; }
 
 public:
+    SharedPtr<BaseTableRef> table_ref_{}; // only used for compact command
```
Comment (JinHai-CN) [inline src/executor/operator/physical_command.cppm:58]: Can we move this memory to CommandInfo?

File: src/function/table/create_index_data.cppm [hunk 3]
```diff
@@ -14,17 +14,20 @@
 
 module;
 
+#include <vector>
```
Comment (JinHai-CN) [inline src/function/table/create_index_data.cppm:17]: can we move the #include out of cppm?

File: src/storage/common/block_index.cppm [hunk 4]
```diff
@@ -25,7 +25,9 @@ struct BlockEntry;
 struct SegmentEntry;
 
 export struct BlockIndex {
-    void Insert(SegmentEntry *segment_entry, TxnTimeStamp timestamp);
+    BlockIndex() = default;
```
Comment (JinHai-CN) [inline src/storage/common/block_index.cppm:28]: Can we remove default constructor?

File: src/storage/meta/entry/segment_column_index_entry.cpp [hunk 5]
```diff
@@ -120,7 +122,189 @@ BufferHandle SegmentColumnIndexEntry::GetIndex() { return vector_buffer_[0]->Loa
...
+                        SizeT block_row_cnt = block_entry->row_count();
+                        segment_column_data.insert(segment_column_data.end(), data_ptr, data_ptr + block_row_cnt * dimension);
+                    }
+                    SizeT total_row_cnt = segment_column_data.size() / dimension;
+                    annivfflat_index->train_centroids(dimension, total_row_cnt, segment_column_data.data());
+                    annivfflat_index->insert_data(dimension, total_row_cnt, segment_column_data.data());
+                    break;
+                }
+                default: {
+                    UnrecoverableError("Not implemented");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_column_index_entry.cpp:161]: This should be RecoverableError. Return Status::NotSupport

File: src/storage/meta/entry/segment_column_index_entry.cpp [hunk 6]
```diff
@@ -120,7 +122,189 @@ BufferHandle SegmentColumnIndexEntry::GetIndex() { return vector_buffer_[0]->Loa
...
+                                    break;
+                                }
+                                case MetricType::kMerticL2: {
+                                    auto hnsw_index =
+                                        static_cast<KnnHnsw<float, u64, PlainStore<float>, PlainL2Dist<float>> *>(buffer_handle.GetDataMut());
+                                    InsertHnsw(hnsw_index);
+                                    break;
+                                }
+                                default: {
+                                    UnrecoverableError("Not implemented");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_column_index_entry.cpp:216]: Error message: Invalid metric type.

File: src/storage/meta/entry/segment_column_index_entry.cpp [hunk 7]
```diff
@@ -120,7 +122,189 @@ BufferHandle SegmentColumnIndexEntry::GetIndex() { return vector_buffer_[0]->Loa
...
+                                }
+                                case MetricType::kMerticL2: {
+                                    auto hnsw_index =
+                                        static_cast<KnnHnsw<float, u64, LVQStore<float, i8, LVQL2Cache<float, i8>>, LVQL2Dist<float, i8>> *>(
+                                            buffer_handle.GetDataMut());
+                                    InsertHnsw(hnsw_index);
+                                    break;
+                                }
+                                default: {
+                                    UnrecoverableError("Not implemented");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_column_index_entry.cpp:239]: Error message: Invalid metric type.

File: src/storage/meta/entry/segment_column_index_entry.cpp [hunk 8]
```diff
@@ -120,7 +122,189 @@ BufferHandle SegmentColumnIndexEntry::GetIndex() { return vector_buffer_[0]->Loa
...
+                                }
+                            }
+                            break;
+                        }
+                        default: {
+                            UnrecoverableError("Not implemented");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_column_index_entry.cpp:245]: Error message: Invalid HNSW encode type.

File: src/storage/meta/entry/segment_column_index_entry.cpp [hunk 9]
```diff
@@ -120,7 +122,189 @@ BufferHandle SegmentColumnIndexEntry::GetIndex() { return vector_buffer_[0]->Loa
...
+                        }
+                    }
+                    break;
+                }
+                default: {
+                    UnrecoverableError("Not implemented");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_column_index_entry.cpp:251]: This should be RecoverableError. Return Status::NotSupport

File: src/executor/operator/physical_command.cppm (hunk 2 above)
Comment (small-turtle-1) [reply src/executor/operator/physical_command.cppm:58]: CommandInfo is defined in parser module. which cannot access `BaseTableRef`

**Review by JinHai-CN — APPROVED** [review]

**Pull request merged**